### PR TITLE
[GFX-949] Properly handle ortho camera in shading_parameters.fs

### PR DESCRIPTION
**Pull request by palver123**
## Jira ticket URL (Why?)
[GFX-949](https://shapr3d.atlassian.net/browse/GFX-949)

## Short description (What? How?) 📖
Filament incorrectly used the ortho camera eye position for obtaining the view vector in shaders. Now it uses the camera forward direction instead. This way you can freely translate the camera along the forward dir (which is good because, you can bring it close to the renderables and you get better numerical precision in eye space calculations)

### Design review 🎨
n/a

### Affected areas 🧭
ortho and perspective camera

### Special use-cases to test 🧷
Fog effect is worth testing in gltfw_viewer, to ensure there was no regression.

### How did you test it? 🤔
Manual 💁‍♂️
Hard to test without Shapr3D. I suggest testing it together with GFX-949_fix_ortho_cam in the shapr3d repository.
I tested the Fog effect in gltf_viewer

Automated 💻
n/a

### CODE REVIEW

**Comment by palver123** [comment]
Review together with https://github.com/shapr3d/shapr3d/pull/8673

File: filament/src/PerViewUniforms.cpp [hunk 1]
```diff
@@ -69,6 +69,7 @@ void PerViewUniforms::prepareCamera(const CameraInfo& camera) noexcept {
     mat4f const& viewFromWorld = camera.view;
     mat4f const& worldFromView = camera.model;
     mat4f const& clipFromView  = camera.projection;
+    const bool isPerspective = clipFromView[2][3] == -1;
```
Comment (veghbalazs) [inline filament/src/PerViewUniforms.cpp:72]: nit: Existing code checks `p[3][2] != 0` to determine projection type ([Camera.cpp#L235](https://github.com/google/filament/blob/afd657d6af0dc7ead4a083685a6e13815e0a9902/filament/src/Camera.cpp#L235)), I'd prefer staying consistent with that.

File: shaders/src/main.fs [hunk 2]
```diff
@@ -40,7 +40,7 @@ void main() {
 #endif
 
 #if defined(HAS_FOG)
-    vec3 view = getWorldPosition() - getWorldCameraPosition();
+    vec3 view = getWorldCameraForward();
```
Comment (veghbalazs) [inline shaders/src/main.fs:43]: Note that `view` should not be a normalized value. This change is not only incorrect in ortho, it breaks perspective. Instead, `view` should be
- in perspective: kept as is, as the vector from eye to fragment position,
- in ortho: the vector from the fragments's orthogonal projection on the near plane to fragment position.

Also `light_punctual.fs` should receive the same treatment.

File: shaders/src/common_getters.fs [hunk 3]
```diff
@@ -43,7 +43,14 @@ highp vec4 getResolution() {
...
 highp vec3 getWorldCameraPosition() {
-    return frameUniforms.cameraPosition;
+    // NOTE: for perspective camera only
+    return frameUniforms.cameraPositionOrForward;
+}
+
+/** @public-api */
+highp vec3 getWorldCameraForward() {
+    // NOTE: for orthographic camera only
+    return frameUniforms.cameraPositionOrForward;
```
Comment (veghbalazs) [inline shaders/src/common_getters.fs:53]: It is quite confusing that only one of `getWorldCameraPosition()` and `getWorldCameraForward()` works, depending on the projection used.

Instead, I suggest introducing a new function in `common_getters.fs`, which may be used to compute unnormalized view vector, and which works in perspective and ortho as well. Since `shading_view/getWorldViewVector()` already returns the direction, a new e.g. `getWorldViewDistance()` could return the magnitude.

File: shaders/src/common_getters.fs [hunk 4]
```diff
@@ -43,7 +43,14 @@ highp vec4 getResolution() {
 
 /** @public-api */
 highp vec3 getWorldCameraPosition() {
-    return frameUniforms.cameraPosition;
```
Comment (veghbalazs) [inline shaders/src/common_getters.fs:46]: Thinking one step further, even though `getWorldCameraPosition()` is only ever used to compute (unnormalized) view vectors in built-in shaders, I would keep it as part of the public shader API.

Of course, there's a contrast that ortho projections are invariant to forward/backward movements, yet any effects that may be built on `getWorldCameraPosition()` are not. This could be resolved by snapping the camera position to the near plane and sending that to shaders, or we can keep it as the responsibility of the client if they indeed use effects mentioned above. (I think we're fine with the latter for our needs).

**Review by veghbalazs — CHANGES_REQUESTED** [review]

File: shaders/src/main.fs (hunk 2 above)
Comment (veghbalazs) [reply shaders/src/main.fs:43]: Nevermind about `light_punctual.fs`, I was on a newer upstream main ([light_punctual.fs#L110](https://github.com/google/filament/blob/afd657d6af0dc7ead4a083685a6e13815e0a9902/shaders/src/light_punctual.fs#L110)), it will have to wait until we catch up with upstream.

File: shaders/src/shading_parameters.fs [hunk 5]
```diff
@@ -33,7 +33,13 @@ void computeShadingParams() {
 #endif
 
     shading_position = vertex_worldPosition;
-    shading_view = normalize(frameUniforms.cameraPosition - shading_position);
+    if (frameUniforms.clipFromViewMatrix[2].w == -1.0) {
```
Comment (marton-shapr3d) [inline shaders/src/shading_parameters.fs:36]: is it OK to test for floating point equality?

File: shaders/src/main.fs (hunk 2 above)
Comment (palver123) [reply shaders/src/main.fs]: Good catch. My idea for detecting such pitfalls is to rename `getWorldCameraPosition()` (e12ac5e85118f0f266f4f52161ee6720ae4b6b26). This way `light_punctual.fs` will fail to compile when we bump to a newer upstream, forcing us to review it.

File: filament/src/PerViewUniforms.cpp (hunk 1 above)
Comment (palver123) [reply filament/src/PerViewUniforms.cpp]: e12ac5e85118f0f266f4f52161ee6720ae4b6b26

File: shaders/src/common_getters.fs (hunk 3 above)
Comment (palver123) [reply shaders/src/common_getters.fs]: I went with introducing `getWorldViewVectorFrom()` (311496e62927a3a9d4753d3c6a31d83eabb3aeb6). For the ortho camera, it is the same as using `getWorldViewVector()` and `getWorldViewDistance()` but for the perspective, there is less calculation. I am up for name suggestions

File: shaders/src/common_getters.fs (hunk 4 above)
Comment (palver123) [reply shaders/src/common_getters.fs:46]: https://github.com/shapr3d/filament/pull/21#discussion_r785707426

File: shaders/src/shading_parameters.fs (hunk 5 above)
Comment (palver123) [reply shaders/src/shading_parameters.fs]: `frameUniforms.clipFromViewMatrix[2].w` will always come from a floating point literal in the C++ code, which [uses IEEE-754 floats](https://en.cppreference.com/w/cpp/language/types#Floating-point_types). Metal uses IEEE-754 too (see [Table 2.1](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf)) and [so does Direct3D 11](https://docs.microsoft.com/en-us/windows/win32/direct3d11/floating-point-rules). So unless `matc` messes up sth it is safe to compare against these literals

File: shaders/src/getters.fs [hunk 6]
```diff
@@ -41,6 +41,18 @@ vec3 getWorldViewVector() {
     return shading_view;
 }
 
+/** @public-api */
+vec3 getWorldViewVectorWithMagnitude() {
+    if (frameUniforms.clipFromViewMatrix[2].w != 0.0) {
```
Comment (gaborvalasek) [inline shaders/src/getters.fs:46]: Nit: I would have written it as
```
    vec3 delta = shading_position - frameUniforms.cameraPosition;
    // Either perspective camera, or distance from plane (orthoEyePos, orthoViewDir)
    return (frameUniforms.clipFromViewMatrix[2].w != 0.0) ? delta : dot(shading_view, delta) * shading_view
```
due to my intense fear of shader compilers doing odd things unless given no alternatives. As in, there's no reason to branch here, so let's try to corner the compiler into generating a conditional move. Mind you, the above code still does not 100% guarantee that but it gives less room to breath.

File: shaders/src/shading_parameters.fs [hunk 7]
```diff
@@ -33,7 +33,13 @@ void computeShadingParams() {
 #endif
 
     shading_position = vertex_worldPosition;
-    shading_view = normalize(frameUniforms.cameraPosition - shading_position);
+    if (frameUniforms.clipFromViewMatrix[2].w != 0.0) {
```
Comment (gaborvalasek) [inline shaders/src/shading_parameters.fs:36]: Nit: similarly to above, I would write this as
```
// Either a perspective, or an orthographic projection
shading_view = (frameUniforms.clipFromViewMatrix[2].w != 0.0) ? frameUniforms.cameraPosition - shading_position : -frameUniforms.cameraForward;
shading_view = normalize(shading_view);
```

**Review by gaborvalasek — APPROVED** [review]

File: docs/Materials.md.html [hunk 8]
```diff
@@ -2131,21 +2131,15 @@
...
-**getUserTimeMode(float m)**        | float    |  Current time modulo m in seconds
-**getExposure()**                   | float    |  Photometric exposure of the camera
-**getEV100()**                      | float    |  [Exposure value at ISO 100](https://en.wikipedia.org/wiki/Exposure_value) of the camera
-
-!!! TIP: world space
-    To achieve good precision, the "world space" in Filament's shading system does not necessarily
-    match the API-level world space. To obtain the position of the API-level camera, custom
-    materials can add `getWorldOffset()` to `getWorldCameraPosition()`.
+                 Name                   |   Type   |            Description
+:---------------------------------------|:--------:|:------------------------------------
```
Comment (marton-shapr3d) [inline docs/Materials.md.html:2135]: nit: why did you change column span?

**Review by marton-shapr3d — APPROVED** [review]

**Review by ViktorHeisenberger — APPROVED** [review]

File: docs/Materials.md.html (hunk 8 above)
Comment (palver123) [reply docs/Materials.md.html:2135]: This is the result of multiple commits, will revert. Thanks for spotting!

File: libs/filamat/src/UibGenerator.cpp [hunk 9]
```diff
@@ -45,13 +45,15 @@ UniformInterfaceBlock const& UibGenerator::getPerViewUib() noexcept  {
             // view
             .add("resolution",              1, UniformInterfaceBlock::Type::FLOAT4, Precision::HIGH)
             // camera
+            .add("cameraForward",           1, UniformInterfaceBlock::Type::FLOAT3, Precision::HIGH)
+            .add("padding0",             1, UniformInterfaceBlock::Type::FLOAT)
```
Comment (marton-shapr3d) [inline libs/filamat/src/UibGenerator.cpp:49]: nit^99: indentation is off

File: shaders/src/shading_parameters.fs (hunk 7 above)
Comment (palver123) [reply shaders/src/shading_parameters.fs]: edff27d8f91b07cc086d9063f7f70fc76e2fc96c

File: shaders/src/getters.fs (hunk 6 above)
Comment (palver123) [reply shaders/src/getters.fs:47]: I kept the if/else because it is widely used in other Filament shaders, plus it is easier to comment the code with the separeted code blocks.

File: libs/filamat/src/UibGenerator.cpp (hunk 9 above)
Comment (palver123) [reply libs/filamat/src/UibGenerator.cpp]: 40fde81a0cc15959dd613d9b8f46d35f12b91f68

File: shaders/src/main.fs (hunk 2 above)
Comment (palver123) [reply shaders/src/main.fs]: Eventually I got rid of `getWorldCameraPosition()` because it was being used incorrectly in Filament (see 63cc63b0b57a3f7c15005351487835f82cd2972e). I updated the usages. In future bumps, `light_punctual.fs` will fail to compile, letting us know it needs our attention.

**Review by veghbalazs — APPROVED** [review]

File: libs/filabridge/include/private/filament/UibStructs.h [hunk 10]
```diff
@@ -54,17 +54,20 @@ struct PerViewUib { // NOLINT(cppcoreguidelines-pro-type-member-init)
 
     math::float4 resolution; // viewport width, height, 1/width, 1/height
 
+    // camera forward in view space (when camera_at_origin is enabled).
+    math::float3 cameraForward;
```
[the remark this answers is not in the document]
Comment (palver123) [reply libs/filabridge/include/private/filament/UibStructs.h:57]: No it doesn't. Good that you point out, because I was about to delete the comment altogether (the whole 'view space' terminology is confusing here and above `cameraPosition` too). Comment deleted in : 5e900c366faaa449afa76c44196b49b0389a1a74

File: samples/sample_full_pbr.cpp [hunk 11]
```diff
@@ -281,9 +281,7 @@ static void setup(Engine* engine, View* view, Scene* scene) {
             vec2 uvDy = dFdy(uv0);
 
             mat3 tangentFromWorld = transpose(getWorldTangentFrame());
-            vec3 tangentCameraPosition = tangentFromWorld * getWorldCameraPosition();
-            vec3 tangentFragPosition = tangentFromWorld * getWorldPosition();
-            vec3 v = normalize(tangentCameraPosition - tangentFragPosition);
+            vec3 v = tangentFromWorld * getWorldViewVector();
```
Comment (palver123) [inline samples/sample_full_pbr.cpp:284]: **For future reference**:
here `tangentFromWorld` is [the usual TBN matrix](https://github.com/shapr3d/filament/blob/shapr3d_master/shaders/src/shading_parameters.fs#L31) containing only rotation and mirroring. Thus it is safe to use it to transform direction vectors. It keeps the vector's length, so the normalizing at the end is not needed (because `getWorldViewVector()` is already normalized)

**Comment by palver123** [comment]
For future reference: these changes don't fix the problem with gradient skybox in ortho view. That is up to [GFX-1045](https://shapr3d.atlassian.net/browse/GFX-1045).
